src/cps/pc_compat/pc_loader.cpp

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: MIT
+// Copyright © 2024 Haowen Liu
+
+#include "cps/pc_compat/pc_loader.hpp"
+
+#include "cps/error.hpp"
+#include "cps/pc_compat/pc.parser.hpp"
+#include "cps/utils.hpp"
+#include "cps/version.hpp"
+
+#include <algorithm>
+#include <filesystem>
+#include <iterator>
+#include <optional>
+#include <ostream>
+#include <stdexcept>
+#include <variant>
+
+#include <fmt/format.h>
+#include <tl/expected.hpp>
+
+namespace cps::pc_compat {
+
+    namespace fs = std::filesystem;
+
+    std::ostream & operator<<(std::ostream & ost, const std::optional<VersionOperation> & version_operation) {
+        if (!version_operation) {
+            return ost;
+        }
+        switch (*version_operation) {
+        case VersionOperation::EQ:
+            ost << "=";
+            break;
+        case VersionOperation::NE:
+            ost << "!=";
+            break;
+        case VersionOperation::LT:
+            ost << "<";
+            break;
+        case VersionOperation::LE:
+            ost << "<=";
+            break;
+        case VersionOperation::GT:
+            ost << ">";
+            break;
+        case VersionOperation::GE:
+            ost << ">=";
+            break;
+        }
+        return ost;
+    }
+
+    std::ostream & operator<<(std::ostream & ost, const PackageRequirement & package_requirement) {
+        ost << package_requirement.package;
+        if (package_requirement.operation) {
+            ost << " " << *package_requirement.operation;
+        }
+        if (package_requirement.version) {
+            ost << " " << *package_requirement.version;
+        }
+        return ost;
+    }
+
+    std::ostream & operator<<(std::ostream & ost, const PackageRequirements & package_requirements) {
+        // Print trailing comma, but it's fine since this is only used for parser debug output
+        for (const auto & package_requirement : package_requirements) {
+            ost << package_requirement << ", ";
+        }
+        return ost;
+    }
+
+    std::ostream & operator<<(std::ostream & ost, const PcPropertyValue & property_value) {
+        if (std::holds_alternative<std::string>(property_value)) {
+            ost << std::get<std::string>(property_value);
+            return ost;
+        }
+        ost << std::get<PackageRequirements>(property_value);
+        return ost;
+    }
+
+    PcLoader::PcLoader() = default;
+
+    tl::expected<loader::Package, std::string> PcLoader::load(std::istream & istream, fs::path const & filename) {
+        scan_begin(istream);
+        yy::parser parse(*this);
+        // To debug parser, uncomment the following line
+        // parse.set_debug_level(true);
```

Comment:
Add a TODO to come back and provide a way to do this without rebuilding cps-config?

Reply:
Yup, makes sense. And there's possibly another item to turn on location tracking and give more intelligible error messages